crates/primitives/src/transaction/mod.rs

```diff
@@ -651,6 +651,108 @@ impl TransactionSigned {
         initial_tx.hash = initial_tx.recalculate_hash();
         initial_tx
     }
+
+    /// Decodes the "raw" format of transaction (e.g. `eth_sendRawTransaction`).
+    ///
+    /// The raw transaction is either a legacy transaction or EIP-2718 type transaction
+    /// For legacy transactions, the format is encoded as: `rlp(tx)`
+    /// For EIP-2718 typed transaction, the format is encoded as the type of the transaction
+    /// followed by the rlp of the transaction: `type` + `rlp(tx)`
+    pub fn decode_enveloped(tx: Bytes) -> Result<Self, DecodeError> {
+        let mut data = tx.as_ref();
+
+        // keep this around so we can use it to calculate the hash
+        let original_encoding = data;
+
+        // decode header
+        let first_header = Header::decode(&mut data)?;
+
+        // Check if its a type byte
+        if data[0] < 0x80 {
+            // decode as envolped type transaction
+            decode_type_transaction(&mut data, first_header)
+        } else {
+            // check if it's a list header
+            if !first_header.list {
+                return Err(DecodeError::Custom("Invalid Input"))
+            }
+            // decode legacy transaction
+            decode_legacy(original_encoding, &mut data, first_header)
+        }
+    }
+}
+
+pub(crate) fn decode_legacy(
+    original_encoding: &[u8],
+    data: &mut &[u8],
+    first_header: Header,
+) -> Result<TransactionSigned, DecodeError> {
```

Comment:
This shouldn't need to have both `original_encoding` and `data`, we should use one argument here